### DISCUSSION TITLE
"undefined reference to `<function>()'"

### DISCUSSION THREAD

**Discussion by voltangle** [corrosion-rs/corrosion#374] [Q&A]
Hi guys! I love that someone actually made something like this, because it removes a lot of problems when integrating Rust in a C/C++ CMake project. Although I managed to make a lot of stuff work, I have a problem, and you have probably already guessed what is it. Linker doesn't find functions in a compiled Rust crate, and I have no idea who and what actually shat the bed and in what direction do I need to look at to fix this issue.

I can't make a reproducible example, so here is the whole project(sorry): https://github.com/ggoraa/edgetx/tree/eldb

But I will also show some relevant files here:
`radio/src/rs/src/lib.rs`:
```rust
#![no_std]
use core::panic::PanicInfo;

mod bindings;

#[panic_handler]
fn halt(info: &PanicInfo) -> ! {
    loop { }
}

#[no_mangle]
pub unsafe extern "C" fn this_function_exists_in_c() -> u32 {
    bindings::cliGetMode()
}
```

`radio/src/rs/CMakeLists.txt`:
```cmake
include(FetchContent)

set(Rust_CARGO_TARGET "thumbv7em-none-eabihf")

FetchContent_Declare(
    Corrosion
    GIT_REPOSITORY https://github.com/corrosion-rs/corrosion.git
    GIT_TAG v0.3.5 # Optionally specify a commit hash, version tag or branch here
)
FetchContent_MakeAvailable(Corrosion)

corrosion_import_crate(MANIFEST_PATH ${RADIO_SRC_DIR}/rs/Cargo.toml NO_STD)
corrosion_set_linker(rs "ld")

include_directories(${RADIO_SRC_DIR}/rs/pub_headers)

target_link_options(firmware PUBLIC "-Wl,--allow-multiple-definition")
target_link_options(firmware PUBLIC "-static")
target_link_libraries(firmware PUBLIC rs)
```

`radio/src/rs/Cargo.toml`:
```toml

[package]
name = "etx-rs"
version = "2.9.0"
edition = "2021"

[lib]
name = "rs"
crate-type = ["staticlib"]
test = false
bench = false

[build-dependencies]
cbindgen = "0.20.0"
bindgen = "0.53.1"

[dependencies]
cty = "0.2"
```

Error itself:
```
<file>: undefined reference to `this_function_exists_in_c()'
collect2: error: ld returned 1 exit status
```

All relevant files are in `radio/src` directory.

To compile, use this command: `cd build && cmake -DPCB=X7 -DCLI=1 -DGCC_ARM_PATH=/Applications/ARM/bin/ -DPYTHON_EXECUTABLE=/usr/local/bin/python3 ..` inside a `build` dir, as usual.

Btw, this project is compiled for STM32, basically ARM Cortex-M

Can you guys help me resolve this issue?

**Comment by jschwe** [comment]
Could you post the complete command you use to **configure** and **build** the project in the devcontainer of the project? 
Is the error message you received the complete error message?

**Reply by voltangle** [reply]
![image](https://user-images.githubusercontent.com/61655029/230589346-943f7995-be54-4809-ab68-648d5cab9a7a.png)
![image](https://user-images.githubusercontent.com/61655029/230589392-a73fbdb0-b8f5-4f2e-9999-7dcfbda73f9d.png)
![image](https://user-images.githubusercontent.com/61655029/230590402-ca096962-3df6-41cb-8ba9-da6cd8338a27.png)

(I am on macOS btw, here is how to do it for other platforms: https://github.com/EdgeTX/edgetx/wiki/Building-and-debugging-EdgeTX)

**Reply by jschwe** [reply]
I can't build the project in the [provided devcontainer](https://github.com/ggoraa/edgetx/blob/eldb/.devcontainer/devcontainer.json), due to missing dependencies.

```
CMake Error at radio/src/lua/debugger/CMakeLists.txt:2 (find_package):
  By not providing "FindNanopb.cmake" in CMAKE_MODULE_PATH this project has
  asked CMake to find a package configuration file provided by "Nanopb", but
  CMake did not find one.

  Could not find a package configuration file provided by "Nanopb" with any
  of the following names:

    NanopbConfig.cmake
    nanopb-config.cmake
```

Could you provide a log where you run the build with `cmake --verbose` (with make it would be `make VERBOSE=1`)? This should print the command line the linker was invoked with.


## Some comments about your CMakeLists.txt

```
corrosion_set_linker(rs "ld")
```

This line is useless for `staticlib`s, since Rust does not invoke the linker for them (on master corrosion issues a warning in such cases). You can just remove it.

```
target_link_options(firmware PUBLIC "-Wl,--allow-multiple-definition")
```

Why do you need this option? You really should not pass this flag unless you have a really good reason and know what you are doing! What are the errors you are seeing when you remove this flag?

```
target_link_options(firmware PUBLIC "-static")
```

Are you sure you need this flag?

**Reply by voltangle** [reply]
First and last flags are here just because I was trying any option I could find to try to make it work. The second flag though is because I was getting errors on multiple definitions, that flag fixed it

> I can't build the project in the [provided devcontainer](https://github.com/ggoraa/edgetx/blob/eldb/.devcontainer/devcontainer.json), due to missing dependencies.

You have probably checked out without submodules, because Nanopb is added using a submodule

**Reply by jschwe** [reply]
> You have probably checked out without submodules, because Nanopb is added using a submodule

Thanks, initializing the submodules helped. 

I got a little further, but still am not at the point you are.

Issue 1: bindgen fails due to libclang not finding `stdint.h`. In response I added `.clang_arg("-I/opt/gcc-arm-none-eabi-10-2020-q4-major/arm-none-eabi/include")`, which helps, but afterwards `stddef.h` is not found (used in cli.h`).  `stddef.h` doesn't seem to be provided by the gcc toolchain in the docker image.

issue 2: commenting the  `stddef.h` include got me further, and the rust library was built successfully. New error:

```
[ 34%] Generating nanopb/generator/proto/nanopb_pb2.py
Traceback (most recent call last):
  File "/workspaces/edgetx/radio/src/thirdparty/nanopb/generator/protoc", line 44, in <module>
    status = invoke_protoc(['protoc'] + args)
  File "/workspaces/edgetx/radio/src/thirdparty/nanopb/generator/proto/_utils.py", line 44, in invoke_protoc
    return subprocess.call(argv)
  File "/usr/lib/python3.6/subprocess.py", line 287, in call
    with Popen(*popenargs, **kwargs) as p:
  File "/usr/lib/python3.6/subprocess.py", line 729, in __init__
    restore_signals, start_new_session)
  File "/usr/lib/python3.6/subprocess.py", line 1364, in _execute_child
    raise child_exception_type(errno_num, err_msg, err_filename)
FileNotFoundError: [Errno 2] No such file or directory: 'protoc': 'protoc'
radio/src/CMakeFiles/firmware.dir/build.make:81: recipe for target 'radio/src/nanopb/generator/proto/nanopb_pb2.py' failed
```

**Reply by voltangle** [reply]
I am really sorry for making you spin up a whole environment for building this :(
It wants a protobuf compiler, `protoc`
It should be just installed using a package manager

Dev containers will also be improved later and work as expected, with all dependencies included

**Comment by jschwe** [comment]
I can reproduce your issue now, but I'm not sure why your linker doesn't see the symbol. 

Regarding the multiple definitions issue - this is an upstream rust issue (https://github.com/rust-lang/rust/issues/104707). In the issue, this [blog post](https://alanwu.space/post/symbol-hygiene/) was linked and describes a way to work around this issue.
